pkg/syndesis/action/upgrade.go

```diff
@@ -0,0 +1,198 @@
+	if a.operatorVersion == "" {
+		operatorVersion, err := version.GetSyndesisVersionFromOperatorTemplate()
+		if err != nil {
+			return err
+		}
+		a.operatorVersion = operatorVersion
+	}
+
+	namespaceVersion, err := version.GetSyndesisVersionFromNamespace(syndesis.Namespace)
+	if err != nil {
+		return err
+	}
+	targetVersion := a.operatorVersion
+
+	resources, err := a.getUpgradeResources(syndesis)
+	if err != nil {
+		return err
+	}
+
+	templateUpgradePod, err := a.findUpgradePod(resources)
+	if err != nil {
+		return err
+	}
+
+	upgradePod, err := a.getUpgradePodFromNamespace(templateUpgradePod, syndesis)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return err
+	}
+
+	if syndesis.Status.ForceUpgrade || k8serrors.IsNotFound(err) {
+		// Upgrade pod not found or upgrade forced
+
+		if namespaceVersion != targetVersion {
+			logrus.Info("Upgrading syndesis resource ", syndesis.Name, " from version ", namespaceVersion, " to ", targetVersion)
+
+			for _, res := range resources {
+				setNamespaceAndOwnerReference(res, syndesis)
+
+				err = createOrReplaceForce(res, true)
+				if err != nil {
+					return err
+				}
+			}
+
+			if syndesis.Status.ForceUpgrade {
+				target := syndesis.DeepCopy()
+				target.Status.ForceUpgrade = false
+
+				return sdk.Update(target)
+			} else {
+				return nil
+			}
+		} else {
+			// No upgrade pod, no version change: upgraded
+			logrus.Info("Syndesis resource ", syndesis.Name, " already upgraded to version ", targetVersion)
+			return upgradeCompleted(syndesis, targetVersion)
+		}
+	} else {
+		// Upgrade pod present, checking the status
+		if upgradePod.Status.Phase == v1.PodSucceeded {
+			// Upgrade finished (correctly)
+
+			// Getting the namespace version again for double check
+			newNamespaceVersion, err := version.GetSyndesisVersionFromNamespace(syndesis.Namespace)
+			if err != nil {
+				return err
+			}
+
+			if newNamespaceVersion == targetVersion {
+				logrus.Info("Syndesis resource ", syndesis.Name, " upgraded to version ", targetVersion)
+				return upgradeCompleted(syndesis, targetVersion)
+			} else {
+				logrus.Warn("Upgrade pod terminated successfully but Syndesis version (", newNamespaceVersion, ") does not reflect target version (", targetVersion, ") for resource ", syndesis.Name, ". Forcing upgrade.")
+				target := syndesis.DeepCopy()
+				target.Status.ForceUpgrade = true
+				target.Status.TargetVersion = targetVersion
+
+				return sdk.Update(target)
+			}
+		} else if upgradePod.Status.Phase == v1.PodFailed {
+			// Upgrade failed
+			logrus.Warn("Failure while upgrading Syndesis resource ", syndesis.Name, " to version ", targetVersion, ": upgrade pod failure")
+
+			target := syndesis.DeepCopy()
+			target.Status.InstallationStatus = v1alpha1.SyndesisInstallationStatusUpgradeFailureBackoff
+			target.Status.Reason = v1alpha1.SyndesisStatusReasonUpgradePodFailed
+			target.Status.Description = "Syndesis upgrade failed (it will be retried again)"
```

Comment:
"Syndesis upgrade from " + namespaceVersion + " to " + targetVersion + " failed ...." ?